Add validation for Authentication and OAuth CRs

stlaz opened the pull request:
This PR adds some basic static validation for the Authentication and OAuth CRs.

/cc @openshift/sig-auth

stlaz commented:
/retest

stlaz commented:
/retest

stlaz commented:
/retest

stlaz commented:
/retest

enj commented on pkg/admission/customresourcevalidation/oauth/validate_oauth.go:
```diff
@@ -108,3 +205,378 @@ func validateOAuthStatus(status configv1.OAuthStatus) field.ErrorList {
+	return errs
+}
+
+func ValidateIdentityProviderName(name string) []string {
```
Use the existing function.

stlaz commented on pkg/admission/customresourcevalidation/authentication/validate_authentication.go:
```diff
@@ -11,19 +11,19 @@ import (
 
 	configv1 "github.com/openshift/api/config/v1"
-	"github.com/openshift/origin/pkg/admission/customresourcevalidation"
+	crvalidation "github.com/openshift/origin/pkg/admission/customresourcevalidation"
```
The lines seemed too long, I tried doing whatever possible to shorten them.

stlaz commented on pkg/admission/customresourcevalidation/authentication/validate_authentication.go:
```diff
@@ -95,16 +94,39 @@ func (authenticationV1) ValidateStatusUpdate(uncastObj runtime.Object, uncastOld
 
 func validateAuthenticationSpec(spec configv1.AuthenticationSpec) field.ErrorList {
```
Split into two which eventually lead to validating the new object, not sure whether we want some more update-specific checks.

stlaz commented:
Pulled in the original test of the Github IdP from `pkg/cmd/server/apis/config/validation` and modified it accordingly.

Also removed the reason string "must be set" from `RequiredField` errors to mirror the original behavior.

ericavonb commented:
/retest

nee1esh commented:
/retest

nee1esh commented:
/retest

ericavonb commented:
/retest

nee1esh commented:
/retest

nee1esh commented:
/retest

sallyom commented on pkg/admission/customresourcevalidation/oauth/validate_github.go:
```diff
@@ -0,0 +1,63 @@
+		errs = append(errs, field.Invalid(fieldPath.Child("teams"), provider.Teams, "specify organizations or teams, not both"))
+	}
+	if len(provider.Teams) == 0 && len(provider.Organizations) == 0 && mappingMethod != configv1.MappingMethodLookup {
+		errs = append(errs, field.Invalid(fieldPath, nil, "no organizations or teams specified, any GitHub user will be allowed to authenticate"))
```
rephrase to sound more like an error, like 'must specify organization or team, or any GH user will be allowed to authenticate'  but btw, what if the goal was to have any GH user authenticate, definitely not ok?

stlaz replied:
Good point, this was previously but a warning, I do not think allowing any GH user is a valid use-case and this would effectively prevent it, but it's a behavior change.

cc @openshift/sig-auth - keep as error/remove and big-fat-document it?

enj replied:
Hmm.  It is definitely valid when using github enterprise.  But we do let you set up gitlab as a "free for all."  We also do not have a good place to give you a warning about this... (I could see a Prometheus alert)

Maybe we start strict and loosen if needed?  Definitely need allow this with GHE.

stlaz replied:
I guess it makes sense to do the same for Google IdP and `hostedDomain`, then.

TODO: still need to allow this for GHE (simple change but haven't done that yet).

stlaz replied:
The same question also stands for all the warnings -> errors in RequestHeader IDP.

ericavonb commented on pkg/admission/customresourcevalidation/oauth/validate_github.go:
```diff
@@ -0,0 +1,63 @@
+
+	if len(provider.Teams) > 0 && len(provider.Organizations) > 0 {
+		errs = append(errs, field.Invalid(fieldPath.Child("organizations"), provider.Organizations, "specify organizations or teams, not both"))
+		errs = append(errs, field.Invalid(fieldPath.Child("teams"), provider.Teams, "specify organizations or teams, not both"))
```
I see that this is coming from the old validation logic, but it doesn't look like there's any reason you _couldn't_ have both orgs and teams. I could see this being useful, like `"organizations": ["keycloak"], "teams": ["openshift/team-red-hat"]`. Maybe that's confusing though?

stlaz replied:
The current oauth-server code would not handle such situations and would fail right away after it noticed that the user is not a part of the `keycloak` organization:

https://github.com/openshift/origin/blob/9c28792c0aece4f9b7926746949111faf61a3be7/pkg/oauthserver/oauth/external/github/github.go#L181-L202

Not sure what lead to such design decision, I bet there is a good reason though, maybe @enj would know.

ericavonb replied:
The user could be a part of the `keycloak` organization. For instance, this would be valid for me: `"organizations": ["quay"], "teams": ["coreos//coreos-full-time"]`

enj replied:
This has never been requested as a "feature" nor do I consider it valuable.  It does not warrant a change to osin and we can always relax the validation later if we ever need it.

stlaz commented on pkg/admission/customresourcevalidation/oauth/validate_oauth.go:
```diff
@@ -93,10 +103,86 @@ func (oauthV1) ValidateStatusUpdate(uncastObj runtime.Object, uncastOldObj runti
+}
+
+func validateOAuthSpecUpdate(newspec, oldspec configv1.OAuthSpec) field.ErrorList {
+	return validateOAuthSpec(newspec)
```
I don't think it needs to

enj replied:
I have been thinking if it makes sense to try and limit how much mutation you can do the the IDPs in one change.  Mostly trying to think if we can help people realize that the name of an IDP cannot be changed.

stlaz commented:
Modified according to changes in master, the OpenShift OAuth API changes and according to https://github.com/openshift/cluster-authentication-operator/pull/99.

stlaz commented on pkg/admission/customresourcevalidation/oauth/validate_oauth.go:
```diff
@@ -93,10 +103,71 @@ func (oauthV1) ValidateStatusUpdate(uncastObj runtime.Object, uncastOldObj runti
+		}
+
+		identityProviderPath := specPath.Child("identityProvider").Index(i)
+		errs := append(errs, ValidateIdentityProvider(identityProvider, identityProviderPath)...)
```
That's a bug, young man!

stlaz commented:
Cleaned the code a bit, added unit tests and check for `AccessTokenMaxAgeSeconds` being negative integer.

ericavonb commented:
/retest

nee1esh commented:
/retest

nee1esh commented:
/retest

stlaz commented on pkg/admission/customresourcevalidation/oauth/validate_openid.go:
```diff
@@ -0,0 +1,36 @@
+
+	allErrs = append(allErrs, ValidateOAuthIdentityProvider(provider.ClientID, provider.ClientSecret, fieldPath.Child("provider"))...)
+
+	if provider.Issuer != strings.TrimRight(provider.Issuer, "/") {
```
@enj I originally implemented due to the request in https://github.com/openshift/cluster-authentication-operator/blob/master/pkg/operator2/idp.go#L237 but that makes this single spot stand out as the single one with this requirement.

Is there a reason why we couldn't continue trimming trailing `/` and adding the oidc discover endpoint instead?

stlaz replied:
Hm, this is actually also a part of check in the request header IdP, I guess we could leave it here then.

stlaz commented:
/hold
Last batch of unit tests is almost ready + there are open questions that need to be answered first by @openshift/sig-auth

stlaz commented:
Added all the unit tests, relaxing the orgs/teams checks for GH ES + UX fixes to the original commit.

stlaz commented:
/retest

nee1esh commented:
/retest

nee1esh commented:
/retest

enj commented:
/lgtm
/hold cancel

enj commented:
Will follow-up in regards to cleaning up any import restriction violations.
